File: android/app/src/main/java/dev/leo/manager/data/Accounts.kt
package dev.leo.manager.data

import kotlinx.serialization.Serializable

/** Coding-agent accounts, as served by /api/accounts. */
@Serializable
data class AccountsView(
    val accounts: List<Account> = emptyList(),
    val signIn: AccountSignIn? = null,
    /** Coding agents whose runs wait for the user to connect, reconnect or resume an account. */
    val required: List<String> = emptyList(),
)

@Serializable
data class AccountWindow(
    val id: String,
    val label: String = "",
    val usedPercent: Double = 0.0,
    val resetsAt: Long? = null,
    val durationMins: Int? = null,
    /** Empty when the window limits every model. */
    val models: List<String> = emptyList(),
) {
    val remaining
        get() = (100.0 - usedPercent).coerceIn(0.0, 100.0)
}

@Serializable data class BankedResets(val available: Int = 0)

@Serializable
data class AccountUsage(
    val windows: List<AccountWindow> = emptyList(),
    val checkedAt: Long? = null,
    val error: String? = null,
    val resets: BankedResets? = null,
) {
    /** Windows limiting every model, shortest first. */
    val general
        get() = windows.filter { it.models.isEmpty() }.sortedBy { it.durationMins ?: 0 }
}

@Serializable
data class Account(
    val id: String,
    val provider: String = "codex",
    val name: String,
    val enabled: Boolean = true,
    val email: String? = null,
    val plan: String? = null,
    val state: String = "pending",
    /** signIn, reconnect, paused, unavailable, waiting, full, next, low or ready. */
    val status: String = "ready",
    val error: String = "",
    val resetError: String = "",
    val maxConcurrentRuns: Int = 4,
    val activeRunIds: List<String> = emptyList(),
    val usage: AccountUsage? = null,
    val remainingPercent: Double? = null,
    val resetsAt: Long? = null,
    val stale: Boolean = true,
)

@Serializable
data class AccountSignIn(
    val accountId: String = "",
    val provider: String = "codex",
    val state: String = "pending",
    val phase: String = "starting",
    val url: String? = null,
    /** A code to enter on OpenAI's page. */
    val code: String? = null,
    /** Whether Anthropic's page shows a code to paste back here. */
    val acceptsCode: Boolean = false,
    val error: String? = null,
)

File: shared/accounts.ts
// Coding-agent accounts, as served by /api/accounts.
export type Provider = 'codex' | 'claude'

/** What the account needs or does, computed by the server's selection rules. */
export type AccountStatus = 'signIn' | 'reconnect' | 'paused' | 'unavailable' | 'waiting' | 'full' | 'next' | 'low' | 'ready'

export interface UsageWindow {
  id: string
  label: string
  usedPercent: number
  resetsAt: number | null
  durationMins: number | null
  /** Empty when the window limits every model. */
  models: string[]
  reached?: boolean
}
export interface Usage {
  allowed: boolean
  windows: UsageWindow[]
  checkedAt: number | null
  error?: string | null
  /** Codex banked resets. */
  resets: { available: number } | null
}
export interface Account {
  id: string
  provider: Provider
  name: string
  enabled: boolean
  email: string | null
  plan: string | null
  state: 'pending' | 'ready' | 'error'
  status: AccountStatus
  error: string
  resetError?: string
  createdAt: number
  checkedAt: number | null
  lastUsedAt: number | null
  maxConcurrentRuns: number
  activeRunIds: string[]
  usage: Usage | null
  remainingPercent: number | null
  resetsAt: number | null
  stale: boolean
  exhausted: boolean
}
export interface SignIn {
  accountId: string
  provider: Provider
  state: 'pending' | 'complete' | 'failed' | 'cancelled'
  phase: 'starting' | 'authorizing' | 'verifying'
  url: string | null
  /** A code to enter on the sign-in page (Codex). */
  code: string | null
  /** Whether the sign-in page shows a code to paste back here (Claude Code). */
  acceptsCode: boolean
  expiresAt: number | null
  error: string | null
}
export interface Accounts {
  accounts: Account[]
  signIn: SignIn | null
  /** Coding agents whose runs wait for the user to connect, reconnect or resume an account. */
  required: Provider[]
}

export const providers: Record<Provider, { label: string, vendor: string, subscription: string }> = {
  codex: { label: 'Codex', vendor: 'OpenAI', subscription: 'ChatGPT account' },
  claude: { label: 'Claude Code', vendor: 'Anthropic', subscription: 'Claude account' },
}

/** Remaining percentage under which a window is running low. */
export const lowPercent = 10

/** The windows limiting every model, shortest first. */
export function generalWindows(usage: Usage | null) {
  return (usage?.windows ?? []).filter(window => !window.models.length).sort((a, b) => (a.durationMins ?? 0) - (b.durationMins ?? 0))
}
export function remainingPercent(window: UsageWindow) {
  return Math.max(0, Math.min(100, 100 - window.usedPercent))
}
